### DISCUSSION TITLE
Running 'git merge-base --fork-point master' command as part of a job fails

### DISCUSSION THREAD

**Discussion by christophebedard** [community/community#26470] [Actions]
<p>Here is my simple config:</p>
<pre><code>name: CI

on:
  pull_request:
  push:
    branches:
      - master

jobs:
  test:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        python-version: [3.6, 3.7, 3.8]

    steps:
    - uses: actions/checkout@v2
      with:
        ref: ${{ github.event.pull_request.head.ref }}
    - name: Fetch all branches
      run: |
        git fetch origin master:master
    - name: Set up Python ${{ matrix.python-version }}
      uses: actions/setup-python@v1
      with:
        python-version: ${{ matrix.python-version }}
    - name: Check
      run: |
        git merge-base --fork-point master
</code></pre>
<p>I’m triggering this on a pull request. The git fetch command is there to actually fetch the master branch (otherwise git merge-base would fail).</p>
<p>However, the git merge-base command still fails and I can’t figure out why.</p>
<pre><code>Check0s
##[error]Process completed with exit code 1.
Run git merge-base --fork-point master
  git merge-base --fork-point master
  shell: /bin/bash -e {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.7.6/x64
##[error]Process completed with exit code 1.
</code></pre>
<p>Other git commands like git branch work fine. But this one just fails without any output and makes the job fail.</p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>Hi <a class="mention" href="/christophebedard">@christophebedard</a> , </p>
<p>First, let me suppose the PR is from dev to master.  Based on my test ‘git merge-base --fork-point master’ command will fail when you have merged dev to master before. </p>
<p>Please see my reproduce steps:</p>
<ol>
<li>
<p>create a new branch dev based on master. </p>
</li>
<li>
<p>make a change in dev and then create a PR</p>
</li>
<li>
<p>this time the workflow run triggered by pull request will succeed</p>
</li>
<li>
<p>merge the PR</p>
</li>
<li>
<p>go to dev branch and then create a new change and a new PR from dev to master</p>
</li>
<li>
<p>the workflow run trigged by this new PR will fail like you reported . </p>
</li>
</ol>
<p>Could you please check whether this scenario matches yours? And please tell me your purpose to run ‘git merge-base --fork-point master’ ? </p>
<p>In addition, the workflow run triggered by <a href="https://help.github.com/en/actions/reference/events-that-trigger-workflows#pull-request-event-pull_request" rel="nofollow noopener">pull request event</a> will get the PR merge branch (refs/pull/:prNumber/merge) as checkout ref by default. </p>

**Comment by christophebedard** [comment]
<blockquote>
<p>Could you please check whether this scenario matches yours?</p>
</blockquote>
<p>Yes, that is exactly my scenario. I thought I tested with merges like this, but I guess not.</p>
<blockquote>
<p>And please tell me your purpose to run ‘git merge-base --fork-point master’ ? </p>
</blockquote>
<p>I’m trying to find the first commit before the change in the PR. In other words: the commit on branch master on which the dev branch is based.</p>
<p>I’m simply trying to get the first and last commits of the PR. The commit that ‘git merge-base --fork-point master’ would return (if it wasn’t using refs/pull/:prNumber/merge) is the one first before the “first” commit, which works as well. In this case, it looks like I can just use ‘pull_request.base.sha’.</p>
<p>Thanks for pointing it out.</p>

**Reply by Frank-Reichenbach** [reply]
The 'github.event.pull_request.base.sha' didn't work for me, but I like to share my solution:
```shell
if [ $GITHUB_BASE_REF ]; then
    # Pull Request
    # 'github.event.before/after' and 'github.event.pull_request.base.sha' are inconsistent for PRs 
    # Find the commit where the PR branch was forked from the base branch
    git fetch origin ${{ github.base_ref }} ${{ github.event.pull_request.head.ref }}
    FORK_POINT=$( git show-branch --merge-base "origin/${{ github.base_ref }}" "origin/${{ github.event.pull_request.head.ref }}" )
    LATEST_COMMIT_IN_PR=${{ github.event.pull_request.head.sha }}
    CHANGED_FILES=$( git diff --name-only $FORK_POINT $LATEST_COMMIT_IN_PR )
    echo "Changes for PR"
    echo "Diff between $FORK_POINT and $LATEST_COMMIT_IN_PR"
else
    # Push
    # 'github.event.before' and 'after' are including all commits of the current push event
    CHANGED_FILES=$(git diff --name-only ${{ github.event.before }} ${{ github.event.after }})
    echo "Changes for PUSH" 
    echo "Diff between ${{ github.event.before }} and ${{ github.event.after }}"
fi
# Print result for verification
echo Changed files:
echo "$CHANGED_FILES"
```
